searx/plugins/__init__.py

```diff
@@ -77,6 +86,60 @@ def call(self, ordered_plugin_list, plugin_type, request, *args, **kwargs):
         return ret
 
 
+def load_external_plugins(plugin_names):
+    plugins = []
+    for name in plugin_names:
+        logger.debug('loading plugin: {0}'.format(name))
+        try:
+            pkg = import_module(name)
+        except Exception as e:
+            logger.critical('failed to load plugin module {0}: {1}'.format(name, e))
+            exit(3)
+
+        pkg.__base_path = dirname(abspath(pkg.__file__))
+
+        fix_package_resources(pkg, name)
+
+        plugins.append(pkg)
+        logger.debug('plugin "{0}" loaded'.format(name))
+    return plugins
+
+
+def check_resource(base_path, resource_path, name, dir_prefix):
```

Comment:
I guess a symbolic link doesn't work ?
The copy can be a problem: if the install folder is read-only, the copy won't work.

https://flask.palletsprojects.com/en/1.1.x/blueprints/#static-files

According to https://stackoverflow.com/questions/9513072/more-than-one-static-path-in-local-flask-instance :

```python
blueprint = Blueprint('site', __name__, static_url_path='/static/site', static_folder='path/to/files')
app.register_blueprint(blueprint)
```

```jinja
{{ url_for('site.static', filename='filename') }}
```

It could even be a way to include templates (I think about answer like "color picker" that display an actual color picker).

Reply:
About blueprint I want to refer my comment: [as far as I can see Theme selection is not possible with Flask-Blueprints. Therefore I first refrain from using blueprints.](https://github.com/asciimoo/searx/pull/1938#issuecomment-618424236)

Reply:
> I guess a symbolic link doesn't work ?
> The copy can be a problem

What would be the advantage of using symlinks?

> the old files won't deleted (if the names change).

@dalf, you are right, it would be better to create a folder for each plugin and check files in the reverse direction too.

> I don't know if blueprint can be a solution ?

Blueprints won't work if you serve your static files with your production webserver - which is recommended in bigger instances.

Reply:
> Blueprints won't work if you serve your static files with your production webserver - which is recommended in bigger instances.

> What would be the advantage of using symlinks?

If searx maintains a static folder made of symbolic links:
* Flask / starlette whatever can serve the files
* nginx / caddy can serve the files too.

In this folder, there is:
* for each plugin, one symbolic link to the static folder.
* (same for the themes ?)
* a symbolic link to the core components.

It is a "temporary" folder outside the searx code base: the git repository / distrib package stays clean.

(Just an idea)

It can be related to https://github.com/asciimoo/searx/issues/2034 : the build process takes all the files, rename with a hash in the name, so the HTTP server can be configure to have a very long expiration date. It works well only if the plugin are included in the process.

Reply:
> It is a "temporary" folder outside the searx code base: the git repository / distrib package stays clean.
> (Just an idea)

I like this idea ..

Reply:
@dalf this sounds good, one disadvantage of symlinks is that you have to configure your webserver to follow them. It isn't a big deal. In case of nginx it is just a `disable_symlinks off;` line in the config, but we have to emphasize it in the documentation at least, however it still can be confusing to somebody who installed searx before this change and tries to add an external plugin.
Also I'm thinking about security impacts of using symlinks. So far I've found these issues:
 - Directory listing with symlinks can be dangerous
 - https://serverfault.com/questions/244592/followsymlinks-on-apache-why-is-it-a-security-risk/244612
 - https://blog.hboeck.de/archives/873-The-tricky-security-issue-with-FollowSymLinks-and-Apache.html
 - https://magento.stackexchange.com/questions/35243/are-symlinks-on-templates-a-security-problem-and-if-yes-why

The last one mentions an interesting issue. It isn't symlink specific, but if a malicious external plugin says that its resource is `/etc/passwd` then we make it accessible from the web. I should check this scenario too.

Copying files feels a bit more straight forward solution, however symlinks could have advantages over file duplication as @dalf mentioned. It's hard to decide.. =) @dalf / devs what is your opinion?

> It is a "temporary" folder outside the searx code base: the git repository / distrib package stays clean.

I'm not sure if I'm right, but as far as I know the symlink also "contaminates" the distro packages.

Reply:
>  but if a malicious external plugin says that its resource is /etc/passwd then we make it accessible from the web. I should check this scenario too.

A malicious external plugin can do everything the searx process can do / there are no special _access controls_ for plugins.

> Also I'm thinking about security impacts of using symlinks .. Copying files feels a bit more straight forward solution, however symlinks could have advantages over file duplication  ...

The more I think about it, I prefer a solution with (only one) `static/` folder where all needed files are _synced_ in.  ATM this folder is our [searx/static](https://github.com/asciimoo/searx/tree/master/searx/static).

This has some drawbacks, e.g. resources from external plugins are copied into the [searx/static](https://github.com/asciimoo/searx/tree/master/searx/static) folder of the repository.

OT: #1938 iterates over [static files and sync them](https://github.com/return42/searx/blob/ep-plugins/searx/resources.py#L111) / triggered by [webapp.py](https://github.com/return42/searx/blob/ep-plugins/searx/webapp.py#L117) calling [get_plugins()](https://github.com/return42/searx/blob/ep-plugins/searx/plugins/__init__.py#L78)

> It can be related to #2034 : the

I vote for a _temporary folder_ which can be loaded (copy files in) by the searx / on the other side a _static-URL_ is needed for this folder under which a HTTP server can export the static files.  The _temporary folder_ is loaded by webapp.py (_writeable_) and _readable_ by the HTTP server process.. The  _temporary folder_ on the file system should be outside of the searx repository.  For developer purpose (and may be the default) we can create the _temporary folder_ in the the git working tree and add them to the `.gitignore`.

Reply:
> A malicious external plugin can do everything the searx process can do / there are no special access controls for plugins.

Thats true, you are right.